tools/templates.js

```diff
@@ -42,16 +44,112 @@ try {
   console.error('Failed to hook Fractal logger', err.stack); // eslint-disable-line no-console
 }
 
+/**
+ * @param {string} s A string in camel case.
+ * @returns {string} The string converted to hyphnated format.
+ */
+const hyphnatedFromCamelCase = s => s.replace(/([A-Z])/g, (match, token) => `-${token.toLowerCase()}`).replace(/^-+/, '');
+
+/**
+ * @param {Object} The key/value pair of attributes.
+ * @returns {string} The attributes formatted as a string.
+ */
+const formatAttributes = attrs => Object.keys(attrs).reduce((acc, key) => `${acc} ${key}="${attrs[key]}"`.trim(), '');
+
+/**
+ * @param {Object} toplevelDescriptor The icon descriptor.
+ * @returns {string} The Handlebars partial string.
+ */
+const js2partial = toplevelDescriptor => {
+  const convert = descriptor => {
+    const { elem = 'svg', attrs = [], formattedAttrs = '', formattedContent = '', content = [] } = descriptor;
+    return `<${elem} ${formatAttributes(attrs)}${formattedAttrs}>${formattedContent}${content.map(convert).join('')}</${elem}>`;
+  };
+  const formattedAttrsFromParams =
+    ' {{#each this}} {{#if (is @key "description")}}aria-label{{else}}{{@key}}{{/if}}="{{this}}"{{/each}}';
+  const formattedAriaLabelAttr = '{{#if description}} aria-label="{{description}}"{{/if}}';
+  return convert({
+    ...toplevelDescriptor,
+    formattedAttrs: formattedAttrsFromParams + formattedAriaLabelAttr,
+    formattedContent: '{{#if description}}<title>{{description}}</title>{{/if}}',
+  });
+};
+
+/**
+ * @param {Object} descriptor The icon descriptor from `carbon-icons` library.
+ * @returns {Object} The icon descriptor in `@carbon/icons` format.
+ */
+const normalizeDescriptor = ({ svgData, width, height, viewBox }) => {
+  const attrs = { width, height, viewBox };
+  return {
+    attrs: Object.keys(attrs)
+      .filter(key => attrs[key])
+      .reduce((o, key) => ({ ...o, [key]: attrs[key] }), {}),
+    content: Object.keys(svgData)
+      .filter(elem => svgData[elem])
+      .reduce(
+        (a, elem) => [
+          ...a,
+          ...svgData[elem].map(data => ({
+            elem: elem.replace(/s$/, ''),
+            attrs: data,
+          })),
+        ],
+        []
+      ),
+  };
+};
+
+/**
+ * @param {boolean} useElements `true` to use one from `carbon-elements`.
+ * @returns {Map<string, string>} A set of icons contents, keyed by hyphnated icon name.
+ */
+const getIconsPartials = useElements => {
```

Comment:
How do we know what the list of available partials are? In other words, how do we know the name of the icon that we want to use.